src/protocol/aggregation/mod.rs

```diff
@@ -52,139 +58,229 @@ where
-
-    let output = aggregate_values(validator.context(), gf2_value_bits).await?;
+    let output =
+        aggregate_values_per_bucket(ctx, converted_value_bits, converted_breakdown_key_bits)
+            .await?;
 
-    validator.validate(vec![output]).await
+    validator.validate(output).await
 }
 
-/// Performs a set of aggregation protocols on binary shared values.
 /// This protocol assumes that devices and/or browsers have applied per-user
 /// capping.
 ///
 /// # Errors
 /// propagates errors from multiplications
-#[tracing::instrument(name = "simple_aggregate_values", skip_all)]
-pub async fn aggregate_values<F, C, S>(
+#[tracing::instrument(name = "aggregate_values_per_bucket", skip_all)]
+pub async fn aggregate_values_per_bucket<F, C, S>(
     ctx: C,
-    contribution_value_bits_gf2: Vec<BitDecomposed<Replicated<Gf2>>>,
-) -> Result<S, Error>
+    contribution_values: Vec<BitDecomposed<S>>,
+    breakdown_keys: Vec<BitDecomposed<S>>,
+) -> Result<Vec<S>, Error>
 where
     F: PrimeField,
     C: UpgradedContext<F, Share = S>,
     S: LinearSecretSharing<F> + BasicProtocols<C, F> + Serializable + 'static,
 {
-    let record_count = contribution_value_bits_gf2.len();
-    let bits = contribution_value_bits_gf2[0].len();
-
-    // mod-convert for later validation
-    let convert_ctx = ctx
-        .narrow(&Step::ConvertValueBits)
-        .set_total_records(record_count);
-    let converted_contribution_values = convert_bits(
-        convert_ctx,
-        stream_iter(contribution_value_bits_gf2),
-        0..u32::try_from(bits).unwrap(),
-    );
+    debug_assert!(contribution_values.len() == breakdown_keys.len());
+    let num_records = contribution_values.len();
+    // for now, we assume that the bucket count is 2^BK::BITS
+    let num_buckets = 1 << breakdown_keys[0].len();
 
-    let aggregate = converted_contribution_values
-        .try_fold(S::ZERO, |mut acc, row| async move {
-            acc += &row.to_additive_sharing_in_large_field();
+    let equality_check_ctx = ctx
+        .narrow(&Step::ComputeEqualityChecks)
+        .set_total_records(num_records);
+    let check_times_value_ctx = ctx
+        .narrow(&Step::CheckTimesValue)
+        .set_total_records(num_records);
+
+    // Generate N streams for each bucket specified by the breakdown key (N = |breakdown_keys|).
+    // A stream is pipeline of contribution values multiplied by the "check bit". A check bit is
+    // a bit that is a share of 1 if the breakdown key matches the bucket, and 0 otherwise.
+    let streams = seq_join(
+        ctx.active_work(),
+        stream_iter(breakdown_keys)
+            .zip(stream_iter(contribution_values))
+            .enumerate()
+            .map(|(i, (bk, v))| {
+                let eq_ctx = &equality_check_ctx;
+                let mul_ctx = &check_times_value_ctx;
+                async move {
+                    let equality_checks = check_everything(eq_ctx.clone(), i, &bk).await?;
```

Comment:
Wait a minute... doesn't "i" refer to the index of the record in the stream of values? Why would we check for equality against that?

Reply:
`i` passed to `check_everything` is used as a record ID. `check_everything` uses the record ID to execute `pregenerate_all_combinations` once per record.

Reply:
Ah, my bad. Makes sense.